### DISCUSSION TITLE
How to use the rotateXyPlane() member function from RankTwoTensorTempl to rotate a vector

### DISCUSSION THREAD

**Discussion by batodon** [idaholab/moose#22761] [Q&A General] [answered]
Hello,

The documentation of [RankTwoTensorTempl](https://mooseframework.inl.gov/docs/doxygen/moose/classRankTwoTensorTempl.html) indicates that `rotateXyPlane()` can rotate a tensor about the Z-axis (x-y-plane). I’m looking to rotate a (unit) vector (i.e., _n) instead. I used the following lines of code below, but none seems to work. 

`rotN_xy = RankTwoTensor::rotateXyPlane(_r);`

`rotN_xy = _n.rotateXyPlane(_r);`

Could someone help me with the appropriate line of code that rotate the vector using the `rotateXyPlane() `function? Thanks.

**Comment by souravmat-git** [comment, marked as answer]
Hi,

I am unsure if the rotateXyPlane() function can rotate a vector.
I will do this in the following way:
```
Real theta = _the_angle_of_rotation;
RankTwoTensor _transformation_matrix;

_transformation_matrix(0, 0) = std::cos(theta);
_transformation_matrix(0, 1) = std::sin(theta);
_transformation_matrix(1, 0) = -std::sin(theta);
_transformation_matrix(1, 1) = std::cos(theta);

RealVectorValue _r;
RealVectorValue rotN_xy = _transformation_matrix * _r;
```

**Reply by batodon** [reply]
Hello! Thanks a million! The code compiles successfully. How about rotation around the X-axis (y-z plane)? I’m assuming I need to change my cosine and sine functions. How do I go about it? Or should I use the same code? Thanks.

**Reply by batodon** [reply]
Hello @souravmat-git , 
For some reason, I don’t know why the previous code compiled successfully because my vector (i.e., `RealVectorValue _r`)  is 3x1. I’m rotating in 3D, not 2D. I made a few changes to the transformation matrix (shown below) and the code compiles (for rotation in the x-y plane). I’m a bit doubtful about my transformation matrix. Could you help? Thanks.  

```
    _transformation_matrix(0, 0) = std::cos(theta);
    _transformation_matrix(0, 1) = std::sin(theta);
    _transformation_matrix(0, 2) = 0;
    _transformation_matrix(1, 0) = -std::sin(theta);
    _transformation_matrix(1, 1) = std::cos(theta);
    _transformation_matrix(1, 2) = 0;
    _transformation_matrix(2, 0) = 0;
    _transformation_matrix(2, 1) = 0;
    _transformation_matrix(2, 2) = 1;
```

**Reply by GiudGiud** [reply]
this looks fine for rotating in the XY plane
see wikipedia they have it written down
https://en.wikipedia.org/wiki/Rotation_matrix

**Reply by batodon** [reply]
sure!. Thanks a lot @GiudGiud.
